ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/CompactionQueryBuilderForInsertOnly.java

```diff
@@ -0,0 +1,313 @@
+   *                       Cannot be null.
+   * @param operation      query's Operation e.g. Operation.CREATE.
+   * @throws IllegalArgumentException if compactionType is null or the compaction type is REBALANCE
+   */
+  CompactionQueryBuilderForInsertOnly(CompactionType compactionType, Operation operation, String resultTableName) {
+    super(compactionType, operation, true, resultTableName);
+    if (CompactionType.REBALANCE.equals(compactionType)) {
+      throw new IllegalArgumentException("Rebalance compaction is supported only on full ACID tables!");
+    }
+  }
+
+  /**
+   * Set the StorageDescriptor of the table or partition to compact.
+   * Required for Create operations in insert-only compaction.
+   *
+   * @param storageDescriptor StorageDescriptor of the table or partition to compact, not null
+   */
+  CompactionQueryBuilder setStorageDescriptor(StorageDescriptor storageDescriptor) {
+    this.storageDescriptor = storageDescriptor;
+    return this;
+  }
+
+  protected void buildSelectClauseForInsert(StringBuilder query) {
+    // Need list of columns for major crud, mmmajor partitioned, mmminor
+    List<FieldSchema> cols;
+    if (CompactionType.MAJOR.equals(compactionType) && sourcePartition != null || CompactionType.MINOR.equals(
+        compactionType)) {
+      if (sourceTab == null) {
+        return; // avoid NPEs, don't throw an exception but skip this part of the query
+      }
+      cols = sourceTab.getSd().getCols();
+    } else {
+      cols = null;
+    }
+    switch (compactionType) {
+    case MAJOR: {
+      if (sourcePartition != null) { //mmmajor and partitioned
+        appendColumns(query, cols, false);
+      } else { // mmmajor and unpartitioned
+        query.append("*");
+      }
+      break;
+    }
+    case MINOR: {
+      appendColumns(query, cols, false);
+    }
+    }
+  }
+
+  protected void getSourceForInsert(StringBuilder query) {
+    if (sourceTabForInsert != null) {
+      query.append(sourceTabForInsert);
+    } else {
+      query.append(sourceTab.getDbName()).append(".").append(sourceTab.getTableName());
+    }
+    query.append(" ");
+    if (CompactionType.MAJOR.equals(compactionType) && StringUtils.isNotBlank(orderByClause)) {
+      query.append(orderByClause);
+    }
+  }
+
+  protected void buildWhereClauseForInsert(StringBuilder query) {
+    if (CompactionType.MAJOR.equals(compactionType) && sourcePartition != null && sourceTab != null) {
+      List<String> vals = sourcePartition.getValues();
+      List<FieldSchema> keys = sourceTab.getPartitionKeys();
+      if (keys.size() != vals.size()) {
+        throw new IllegalStateException("source partition values (" + Arrays.toString(
+            vals.toArray()) + ") do not match source table values (" + Arrays.toString(
+            keys.toArray()) + "). Failing compaction.");
+      }
+
+      query.append(" where ");
+      for (int i = 0; i < keys.size(); ++i) {
```

Comment:
keys.size() is being computed twice, here & above
```
 if (keys.size() != vals.size()) {
```
can refactor into a variable and use at both places